src/gentropy/datasource/biosample_ontologies/utils.py

```diff
@@ -0,0 +1,197 @@
+        explode_outer("graph.edges").alias("edge")
+    ).select(
+        col("edge.sub").alias("subject"),
+        col("edge.pred").alias("predicate"),
+        col("edge.obj").alias("object")
+    )
+    df_edges = df_edges.withColumn("subject", regexp_replace(col("subject"), "http://purl.obolibrary.org/obo/", ""))
+    df_edges = df_edges.withColumn("object", regexp_replace(col("object"), "http://purl.obolibrary.org/obo/", ""))
+
+    # Extract the relevant information from the nodes
+    transformed_df = df_nodes.select(
+    regexp_replace(col("node.id"), "http://purl.obolibrary.org/obo/", "").alias("biosampleId"),
+    col("node.lbl").alias("biosampleName"),
+    col("node.meta.definition.val").alias("description"),
+    collect_set(col("node.meta.xrefs.val")).over(Window.partitionBy("node.id")).getItem(0).alias("xrefs"),
+    # col("node.meta.deprecated").alias("deprecated"),
+    collect_set(col("node.meta.synonyms.val")).over(Window.partitionBy("node.id")).getItem(0).alias("synonyms"))
+
+
+    # Extract the relationships from the edges
+    # Prepare relationship-specific DataFrames
+    df_parents = df_edges.filter(col("predicate") == "is_a").select("subject", "object").withColumnRenamed("object", "parent")
+    df_children = df_edges.filter(col("predicate") == "is_a").select("object", "subject").withColumnRenamed("subject", "child")
+
+    # Aggregate relationships back to nodes
+    df_parents_grouped = df_parents.groupBy("subject").agg(array_distinct(collect_list("parent")).alias("parents"))
+    df_children_grouped = df_children.groupBy("object").agg(array_distinct(collect_list("child")).alias("children"))
+
+    # Get all ancestors
+    df_with_ancestors = json_graph_traversal(df_parents_grouped, "subject", "parents", "ancestors")
+    # Get all descendants
+    df_with_descendants = json_graph_traversal(df_children_grouped, "object", "children", "descendants")
+
+    # Join the ancestor and descendant DataFrames
+    df_with_relationships = df_with_ancestors.join(df_with_descendants, df_with_ancestors.subject == df_with_descendants.object, "full_outer").withColumn("biosampleId", coalesce(df_with_ancestors.subject, df_with_descendants.object)).drop("subject", "object")
+
+    # Join the original DataFrame with the relationship DataFrame
+    final_df = transformed_df.join(df_with_relationships, ["biosampleId"], "left")
+
+    return BiosampleIndex(
+        _df=final_df,
+        _schema=BiosampleIndex.get_schema()
+        )
+
+def merge_biosample_indices(
+    biosample_indices : list[BiosampleIndex]
+    ) -> BiosampleIndex:
+    """Merge a list of biosample indices into a single biosample index.
+
+    Where there are conflicts, in single values - the first value is taken. In list values, the union of all values is taken.
+
+    Args:
+        biosample_indices (list[BiosampleIndex]): Biosample indices to merge.
+
+    Returns:
+        BiosampleIndex: Merged biosample index.
+    """
+
+    def merge_lists(
```

Comment:
It's a great function, nice elegant logic! A few comments:

- I would add a description explaining that the order of element are not kept (lists are ordered structures). 
- It returns with a unique set of elements, instead of all elements.
- I would put this function under `common.utils.py`, because such function might be useful in other places. 

Question: Should this function be generalised to flatten arbitrarily deeply nested lists? (via recursion?)